test/test_rules_patterns.py

```diff
@@ -1,7 +1,133 @@
+        (
+            "MatchQ[x, F[x1_:0, F[x2_:0,y_]]]",
+            "True",
+            "With OneIdentity, pattern nested",
+        ),
+        (
+            "MatchQ[x, G[x1_:0, G[x2_:0,y_]]]",
+            "False",
+            "With OneIdentity, pattern nested",
+        ),
+        (
+            "MatchQ[x, F[x1___:0, F[x2_:0,y_]]]",
+            "True",
+            "With OneIdentity, pattern nested",
+        ),
+        (
+            "MatchQ[x, G[x1___:0, G[x2_:0,y_]]]",
+            "False",
+            "With OneIdentity, pattern nested",
+        ),
+        ("MatchQ[x, F[F[x2_:0,y_],x1_:0]]", "True", "With OneIdentity, pattern nested"),
+        (
+            "MatchQ[x, G[G[x2_:0,y_],x1_:0]]",
+            "False",
+            "With OneIdentity, pattern nested",
+        ),
+        ("MatchQ[x, F[x_.,y_]]", "False", "With OneIdentity, and Optional, no default"),
+        (
+            "MatchQ[x, G[x_.,y_]]",
+            "False",
+            "Without OneIdentity, and Optional, no default",
+        ),
+        ("Default[F, 1]=1.", "1.", None),
+        ("Default[G, 1]=2.", "2.", None),
+        ("MatchQ[x, F[x_.,y_]]", "True", "With OneIdentity, and Optional, default"),
+        ("MatchQ[x, G[x_.,y_]]", "False", "Without OneIdentity, and Optional, default"),
+        ("MatchQ[F[F[H[y]]],F[x_:0,u_H]]", "False", None),
+        ("MatchQ[G[G[H[y]]],G[x_:0,u_H]]", "False", None),
+        ("MatchQ[F[p, F[p, H[y]]],F[x_:0,u_H]]", "False", None),
+        ("MatchQ[G[p, G[p, H[y]]],G[x_:0,u_H]]", "False", None),
+        (None, None, None),
+    ],
+)
+def test_one_identity(str_expr, str_expected, msg):
+    check_evaluation(
+        str_expr,
+        str_expected,
+        to_string_expr=True,
+        to_string_expected=True,
+        hold_expected=True,
+        failure_message=msg,
+    )
+
+
+@pytest.mark.parametrize(
+    ("str_expr", "str_expected", "msg"),
+    [
+        (None, None, None),
+        # F has the attribute, but G doesn't.
+        ("SetAttributes[F, OneIdentity]", "Null", None),
+        ("SetAttributes[r, Flat]", "Null", None),
+        ("SetAttributes[s, Flat]", "Null", None),
+        ("SetAttributes[s, OneIdentity]", "Null", None),
+        # Replace also takes into account the OneIdentity attribute,
+        # and also modifies the interpretation of the Flat attribute.
+        (
+            "F[a,b,b,c]/.F[x_,x_]->Fp[x]",
+            "F[a, b, b, c]",
+            "https://reference.wolfram.com/language/tutorial/Patterns.html",
+        ),
+        (
+            "r[a,b,b,c]/.r[x_,x_]->rp[x]",
+            "r[a, rp[r[b]], c]",
+            "https://reference.wolfram.com/language/tutorial/Patterns.html",
+        ),
+        (
+            "s[a,b,b,c]/.s[x_,x_]->sp[x]",
+            "s[a, rp[b], c]",
+            "https://reference.wolfram.com/language/tutorial/Patterns.html",
+        ),
+        (None, None, None),
+    ],
+)
```

Comment:
Thanks for adding these tests. The ones from Patterns.html I think should go in a doctests with suitable commentary; they were written with the intention of describing how `OneIdentity` works.